File: ios/Packages/SolennixCore/Sources/SolennixCore/Cache/CachedProduct.swift
import Foundation
import SwiftData

// MARK: - CachedProduct

/// SwiftData model mirroring the `Product` struct for offline caching.
@Model
public final class CachedProduct {

    // MARK: - Properties

    @Attribute(.unique)
    public var id: String

    public var userId: String
    public var name: String
    public var category: String
    public var basePrice: Double
    /// JSON string representation of the recipe `AnyCodable` value.
    public var recipeJSON: String?
    public var imageUrl: String?
    public var isActive: Bool
    public var createdAt: String
    public var updatedAt: String

    // MARK: - Init

    public init(
        id: String,
        userId: String,
        name: String,
        category: String,
        basePrice: Double,
        recipeJSON: String? = nil,
        imageUrl: String? = nil,
        isActive: Bool,
        createdAt: String,
        updatedAt: String
    ) {
        self.id = id
        self.userId = userId
        self.name = name
        self.category = category
        self.basePrice = basePrice
        self.recipeJSON = recipeJSON
        self.imageUrl = imageUrl
        self.isActive = isActive
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    /// Creates a cached version from a `Product` struct.
    public convenience init(from product: Product) {
        var recipeString: String?
        if let recipe = product.recipe {
            if let data = try? JSONEncoder().encode(recipe) {
                recipeString = String(data: data, encoding: .utf8)
            }
        }

        self.init(
            id: product.id,
            userId: product.userId,
            name: product.name,
            category: product.category,
            basePrice: product.basePrice,
            recipeJSON: recipeString,
            imageUrl: product.imageUrl,
            isActive: product.isActive,
            createdAt: product.createdAt,
            updatedAt: product.updatedAt
        )
    }

    // MARK: - Conversion

    /// Converts this cached model back to a `Product` value type.
    public func toProduct() -> Product {
        var recipe: AnyCodable?
        if let json = recipeJSON, let data = json.data(using: .utf8) {
            recipe = try? JSONDecoder().decode(AnyCodable.self, from: data)
        }

        return Product(
            id: id,
            userId: userId,
            name: name,
            category: category,
            basePrice: basePrice,
            recipe: recipe,
            imageUrl: imageUrl,
            isActive: isActive,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

File: android/core/database/src/main/java/com/creapolis/solennix/core/database/entity/CachedProduct.kt
package com.creapolis.solennix.core.database.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.creapolis.solennix.core.model.Product

@Entity(tableName = "products")
data class CachedProduct(
    @PrimaryKey val id: String,
    @ColumnInfo(name = "user_id") val userId: String,
    val name: String,
    val category: String,
    @ColumnInfo(name = "base_price") val basePrice: Double,
    val recipe: String?,
    @ColumnInfo(name = "image_url") val imageUrl: String?,
    @ColumnInfo(name = "is_active") val isActive: Boolean,
    @ColumnInfo(name = "staff_team_id") val staffTeamId: String?,
    @ColumnInfo(name = "created_at") val createdAt: String,
    @ColumnInfo(name = "updated_at") val updatedAt: String
)

fun CachedProduct.asExternalModel() = Product(
    id = id,
    userId = userId,
    name = name,
    category = category,
    basePrice = basePrice,
    recipe = recipe,
    imageUrl = imageUrl,
    isActive = isActive,
    staffTeamId = staffTeamId,
    createdAt = createdAt,
    updatedAt = updatedAt
)

fun Product.asEntity() = CachedProduct(
    id = id,
    userId = userId,
    name = name,
    category = category,
    basePrice = basePrice,
    recipe = recipe,
    imageUrl = imageUrl,
    isActive = isActive,
    staffTeamId = staffTeamId,
    createdAt = createdAt,
    updatedAt = updatedAt
)
